knpu/container.md

```diff
@@ -0,0 +1,217 @@
+# Dependency Injection Container
+
+Our project now has services, an interface, and is fully using dependency
+injection. Nice work! One of the downsides of DI is that all the complexity
+of creating and configuring objects is now your job. This isn't so bad since
+it all happens in one place and gives you so much control, but it is something
+we can improve!
+
+If you want to make this easier, the tool you need is called a dependency injection
+container. A lot of DI containers exist in PHP, but let's use Composer to grab the
+simplest one of all, called [Pimple][1]. Add a `require` key to `composer.json` to
+include the library:
+
+[[[ code('002395a9a7') ]]]
+
+Make sure you've [downloaded Composer][2], and then run `php composer.phar install`
+to download Pimple.
+
+***SEEALSE
+If you're new to Composer, check out our free [The Wonderful World of Composer Tutorial][3].
+***
+
+Pimple is both powerful, and tiny. Kind of like having one on prom night. 
+It is just a single file taking up around 200 lines. That's one reason I love it!
+
+Create a new Pimple container. This is an object of course, but it looks
+and acts like an array that we store all of our service objects on:
+
+[[[ code('d7531b4518') ]]]
+
+Start by adding the `SmtpMailer` object under a key called `mailer`. Instead
+of setting it directly, wrap it in a call to `share()` and in an anonymous
+function. We'll talk more about this in a second, but just return the mailer
+object from the function for now:
+
+[[[ code('bc06a6f9b3') ]]]
+
+To access the `SmtpMailer` object, use the array syntax again:
+
+[[[ code('a2f009ed4c') ]]]
+
+It's that simple! Run the application to spam... I mean send great opportunities
+to our friends!
+
+```bash
+php app.php
+```
+
+## Shared and Lazy Services
+
+We haven't fully seen the awesomeness of the container yet, but there are
+already some cool things happening. First, wrapping the instantiation of
+the `mailer` service in an anonymous function makes its creation "lazy":
+
+[[[ code('1f1a463998') ]]]
+
+This means that the object isn't created until much later when we reference
+the `mailer` service and ask the container to give it to us. And if we
+never reference `mailer`, it's never created at all - saving us time and
+memory.
+
+Second, using the `share()` method means that no matter how many times we
+ask for the `mailer` service, it only creates it once. Each call returns 
+the original object:
+
+```php
+    $mailer1 = $container['mailer'];
+    $mailer2 = $container['mailer'];
+
+    // there is only 1 mailer, the 2 variables hold the same one
+    $willBeTrue = $mailer1 === $mailer2;
+```
```

Comment:
... and of course, we need to activate syntax highlight still - KnpLabs/KnpUniversity#414